File: 526-Beautiful Arrangement.java
//Time Complexity=O(N!)
//Space Complexity=O(N)


class Solution {
    int count;
    public int countArrangement(int N) {
        int []arr=new int[N+1];
        if(N==0) return 0;
        for(int i=1;i<=N;i++)
        {
            arr[i]=i;
        }
        count=0;
        helper(arr,1);
        return count;
    }
    private void helper(int []arr,int index)
    {
        if(index==arr.length)
        {
            count++;
            return;
        }
        for(int i=index;i<arr.length;i++)
        {
            if(index%arr[i]==0 || arr[i]%index==0)//Initially we are checking the conditions whether the digit is divisible by index or vice-versa
            {
                //Action
                swap(arr,index,i);//Here we are swapping the value and incrementing the index
                //Recurse
                helper(arr,index+1);
                //Backtrack
                swap(arr,index,i);
            }
        }
    }
    private void swap(int []arr,int i,int j)
    {
        int temp=arr[i];
        arr[i]=arr[j];
        arr[j]=temp;
    }
}

File: 526. Beautiful Arrangement.cpp
// Time Complexity : O(Exponential)
// Space Complexity : O(Exponential)
//The set data structure is not time efficient try to use vector or array if possible.
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
class Solution {
public:
    int count=0;
    int countArrangement(int n) {
        int nums[16] = {0};
        helper(n, 1, nums);
        return count;
    }
    void helper(int n, int index, int nums[]){
        //base case
        if(index>n){
            count++;
            return;
        }
        //logic
        for(int i=1;i<=n;i++){
            if(nums[i]==0  && ((index%i)==0 || (i%index)==0)){
                //action
                nums[i]=i;
                //recurse
                helper(n,index+1,nums);
                //backtrack
                nums[i]=0;
            }
        }
    }
};
